whois-nrtm4/src/main/java/net/ripe/db/nrtm4/SnapshotFileGenerator.java

```diff
@@ -0,0 +1,116 @@
+
+    private final DeltaTransformer deltaTransformer;
+    private final NrtmVersionInfoRepository nrtmVersionInfoRepository;
+    private final SnapshotInitializer snapshotInitializer;
+    private final SnapshotFileStreamer snapshotFileStreamer;
+    private final SnapshotFileRepository snapshotFileRepository;
+    private final SnapshotSynchronizer snapshotSynchronizer;
+    private final NrtmFileUtil nrtmFileUtil;
+    private final SerialDao serialDao;
+
+    public SnapshotFileGenerator(
+        final DeltaTransformer deltaTransformer,
+        final NrtmVersionInfoRepository nrtmVersionInfoRepository,
+        final SnapshotInitializer snapshotInitializer,
+        final SnapshotFileStreamer snapshotFileStreamer,
+        final SnapshotFileRepository snapshotFileRepository,
+        final SnapshotSynchronizer snapshotSynchronizer,
+        final NrtmFileUtil nrtmFileUtil,
+        final SerialDao serialDao
+    ) {
+        this.deltaTransformer = deltaTransformer;
+        this.nrtmVersionInfoRepository = nrtmVersionInfoRepository;
+        this.snapshotInitializer = snapshotInitializer;
+        this.snapshotFileStreamer = snapshotFileStreamer;
+        this.snapshotFileRepository = snapshotFileRepository;
+        this.snapshotSynchronizer = snapshotSynchronizer;
+        this.nrtmFileUtil = nrtmFileUtil;
+        this.serialDao = serialDao;
+    }
+
+    // TODO: Add a global lock to ensure that no other instance can run until this method exits
+    public Optional<PublishableSnapshotFile> createSnapshot(final NrtmSource source) {
+
+        // Get last version from database.
+        final Optional<NrtmVersionInfo> lastVersion = nrtmVersionInfoRepository.findLastVersion(source);
+        NrtmVersionInfo version;
+        if (lastVersion.isEmpty()) {
+            version = snapshotInitializer.init(source);
+        } else {
+            version = lastVersion.get();
+            if (version.getType() == NrtmDocumentType.DELTA) {
+                version = nrtmVersionInfoRepository.copyAsSnapshotVersion(version);
+            } else {
+                LOGGER.info("Not generating snapshot file since no deltas have been published since v{} with serialID {}",
+                    version.getVersion(), version.getLastSerialId());
+                return Optional.empty();
+            }
+        }
+        // TODO: apply pending deltas
+        //       * if snapshot version > 1...
+        //           - find list of deltas since the last snapshot
+        //           - process them with SnapshotSynchronizer to bring snapshot_objects up to date
+        if (version.getVersion() > 1) {
+            final NrtmVersionInfo lastSnapshot = nrtmVersionInfoRepository.findLastSnapshotVersion(source);
+            final List<SerialEntry> whoisChanges = serialDao.getSerialEntriesBetween(lastSnapshot.getLastSerialId(), version.getLastSerialId());
+            final List<DeltaChange> deltas = deltaTransformer.toDeltaChange(whoisChanges);
```

Comment:
This looks like job of snapshotSynchronizer? As Initiliazor is doing all the work reletd to serialDoa ? This way we do noit need to inject SErialDoa bean here ?

Reply:
yes, done